src/YawPitchControl/DeviceSensor.js

```diff
@@ -0,0 +1,174 @@
+import Component from "@egjs/component";
+import {RelativeOrientationSensor} from "motion-sensors-polyfill";
+import {vec3, glMatrix, quat} from "gl-matrix";
+import {getDeltaYaw, getDeltaPitch} from "./utils";
+import {util as mathUtil} from "../utils/math-util";
+
+const X_AXIS_VECTOR = vec3.fromValues(1, 0, 0);
+const Y_AXIS_VECTOR = vec3.fromValues(0, 1, 0);
+// Quaternion to rotate from sensor coordinates to WebVR coordinates
+const SENSOR_TO_VR = quat.setAxisAngle(quat.create(), X_AXIS_VECTOR, -Math.PI / 2);
+
+export default class DeviceSensor extends Component {
+	constructor() {
+		super();
+
+		this._enabled = false;
+		this._calibrated = false;
+
+		this._sensor = new RelativeOrientationSensor({
+			frequency: 60,
+			coordinateSystem: "screen", // for polyfill
+			referenceFrame: "screen"
+		});
+
+		this._prevQuaternion = null;
+	}
+
+	enable() {
+		if (this._enabled) {
+			return Promise.resolve("Sensor already enabled");
+		}
+		if (!navigator || !navigator.permissions) {
+			// iOS
+			this._startSensor();
+			return Promise.resolve();
+		}
+
+		return Promise.all([
+			navigator.permissions.query({name: "accelerometer"}),
+			navigator.permissions.query({name: "gyroscope"})
+		]).then(results => {
+			if (results.every(result => result.state === "granted")) {
+				this._startSensor();
+			}
+		}).catch(() => {
+			// Start it anyway, workaround for Firefox
+			this._startSensor();
+		});
+	}
+
+	disable() {
+		if (!this._enabled) {
+			return;
+		}
+
+		this._prevQuaternion = null;
+		this._sensor.removeEventListener("read", this._onSensorRead);
+		this._sensor.stop();
+	}
+
+	isEnabled() {
+		return this._enabled;
+	}
+
+	getYawPitchDelta() {
+		const prevQuat = this._prevQuaternion;
+		const currentQuat = this._getOrientation();
+		const defaultYawPitchDelta = {
+			yaw: 0,
+			pitch: 0,
+		};
+
+		if (!currentQuat) {
```

Comment:
@WoodNeck 

Is there a lack of currentQuat? I think it always exists.

Reply:
Good point.